### ISSUE TITLE
Migrate to React Native's New Architecture

### ISSUE THREAD

**Issue by dcalhoun** [WordPress/gutenberg#53360]
This issues tracks various tasks to upgrade third-party dependencies for the native mobile editor. The following tasks could likely be addressed individually as miniature projects.

## What problem does this address?

React Native’s [New Architecture](https://reactnative.dev/docs/the-new-architecture/landing-page) could provide performance benefits (e.g. [JSI](https://reactnative.dev/docs/the-new-architecture/why#new-architectures-improvements)) and may eventually be required for future React Native upgrades or compatibility with third-party dependencies.

## What is your proposed solution?

Some prerequisite items may already have been completed (but have not been verified as complete), while some items may require a significant amount of work. If you can verify that an item is complete or would like to tackle an item, feel free to assign yourself and mark it off the list.

### Prerequisites for Supporting the New Architecture in JavaScript

- [ ] [Define Specs in JavaScript](https://reactnative.dev/docs/new-architecture-library-intro#define-specs-in-javascript)
- [ ] [Ensure autolinking is enabled](https://reactnative.dev/docs/new-architecture-library-intro#make-sure-autolinking-is-enabled)
- [ ] [Configure Codegen](https://reactnative.dev/docs/new-architecture-library-intro#configure-codegen)
- [ ] [Migrate from UIManager JavaScript APIs](https://reactnative.dev/docs/new-architecture-library-intro#migrating-from-uimanager-javascript-apis)

### Enabling in an Android Library

- [ ] [Extend or implement the code-generated Native Interfaces](https://reactnative.dev/docs/new-architecture-library-android#1-extend-or-implement-the-code-generated-native-interfaces)

### Enabling in an iOS Library

- [ ] [Update Podspec for the New Architecture](https://reactnative.dev/docs/new-architecture-library-ios#1-updating-your-podspec-for-the-new-architecture)
- [ ] [Extend or implement the code-generated Native Interfaces](https://reactnative.dev/docs/new-architecture-library-ios#2-extend-or-implement-the-code-generated-native-interfaces)

### Prerequisites for WPiOS

- [x] [Upgrade to React Native version 0.71.0](https://reactnative.dev/docs/new-architecture-app-intro#update-to-the-latest-react-native-version)
- [x] [Use Hermes as an xcframework](https://reactnative.dev/docs/new-architecture-app-intro#use-hermes)
- [ ] [Build the project](https://reactnative.dev/docs/new-architecture-app-intro#ios---build-the-project)
- [ ] [Use Objective-C++ (.mm extension)](https://reactnative.dev/docs/new-architecture-app-intro#ios---use-objective-c-mm-extension)
- [ ] [Make AppDelegate conform to RCTAppDelegate (or custom bridged equivalent)](https://reactnative.dev/docs/new-architecture-app-intro#ios---make-your-appdelegate-conform-to-rctappdelegate)
- [ ] [Run pod install and apply changes](https://reactnative.dev/docs/new-architecture-app-intro#ios---run-pod-install)

### Prerequisites for WPAndroid

- [x] [Upgrade to React Native version 0.71.0](https://reactnative.dev/docs/new-architecture-app-intro#update-to-the-latest-react-native-version)
- [x] [Use Hermes](https://reactnative.dev/docs/new-architecture-app-intro#use-hermes)

### Third-Party Dependencies

Each third-party dependency will need to be upgraded to a version with New Architecture support or replaced entirely in this project.

- [ ] Address `react-native-get-random-values`
- [x] Address `react-native-safe-area-context`
- [x] Address `react-native-screens`
- [x] Address `react-native-svg`
- [x] Address `react-native-webview`
- [ ] Address `@react-native-masked-view/masked-view`
- [ ] Address `@react-native-clipboard/clipboard`
- [x] Address `react-native-reanimated`
- [x] Address `react-native-gesture-handler`
- [ ] Address `react-native-linear-gradient`
- [x] Address `react-native-prompt-android`
- [ ] Address `react-native-video`
- [ ] Address `@react-native-community/slider`
- [ ] Address `react-native-safe-area`
- [ ] Address `react-native-fast-image`

The following table represents the status of each third-party dependency's support for the new React Native architecture.

| Dependency                                                                                                   | Support Status         | Notes                                                                                                                                               |
| ------------------------------------------------------------------------------------------------------------ | -------------- | --------------------------------------------------------------------------------------------------------------------------------------------------- |
| [react-native-get-random-values](https://github.com/LinusU/react-native-get-random-values)                   | 🔴 Not Started | [Issue](https://github.com/LinusU/react-native-get-random-values/issues/38), [PR](https://github.com/LinusU/react-native-get-random-values/pull/40) |
| [react-native-safe-area-context](https://github.com/th3rdwave/react-native-safe-area-context)                | 🟢 Completed   | [v4.1.2](https://github.com/th3rdwave/react-native-safe-area-context/releases/tag/v4.1.2)                                                           |
| [react-native-screens](https://github.com/software-mansion/react-native-screens)                             | 🟢 Completed   | [3.12.0](https://github.com/software-mansion/react-native-screens/releases/tag/3.12.0)                                                              |
| [react-native-svg](https://github.com/software-mansion/react-native-svg)                                     | 🟢 Completed   | [v13.0.0](https://github.com/react-native-svg/react-native-svg/releases/tag/v13.0.0)                                                                |
| [react-native-webview](https://github.com/react-native-webview/react-native-webview)                         | 🟢 Completed   | [v12.0.0](https://github.com/react-native-webview/react-native-webview/releases/tag/v12.0.0)                                                        |
| [@react-native-masked-view/masked-view](https://github.com/react-native-masked-view/masked-view)             | 🟡 In Progress | [Issue](https://github.com/react-native-masked-view/masked-view/issues/162)                                                                         |
| [@react-native-clipboard/clipboard](https://github.com/react-native-clipboard/clipboard)                     | 🟢 Completed | [v1.13.0](https://github.com/react-native-clipboard/clipboard/releases/tag/v1.13.0)         |
| [react-native-reanimated](https://github.com/software-mansion/react-native-reanimated)                       | 🟢 Completed   | [3.0.0](https://github.com/software-mansion/react-native-reanimated/releases/tag/3.0.0-rc.0)                                                        |
| [react-native-gesture-handler](https://github.com/software-mansion/react-native-gesture-handler)             | 🟢 Completed   | [2.3.0](https://github.com/software-mansion/react-native-gesture-handler/releases/tag/2.3.0)                                                        |
| [react-native-linear-gradient](https://github.com/react-native-linear-gradient/react-native-linear-gradient) | 🟢 Completed   | [Forked](https://github.com/FyndX/react-native-gradient)                                                                                            |
| [react-native-prompt-android](https://github.com/shimohq/react-native-prompt-android)                        | Removed |                                                                                                                                            |
| [react-native-video](https://github.com/react-native-video/react-native-video)                               | 🔴 Not Started | [No lead, blocked by v6](https://github.com/react-native-video/react-native-video/issues/2650)                                                      |
| [@react-native-community/slider](https://github.com/callstack/react-native-slider)                           | 🟢 Completed   | [v4.3.0](https://github.com/callstack/react-native-slider/releases/tag/v4.3.0)                                                                      |
| [react-native-safe-area](https://github.com/miyabi/react-native-safe-area#readme)                            | 🔴 Not Started | Inactive                                                                                                                                            |
| [react-native-fast-image](https://github.com/DylanVann/react-native-fast-image)                              | 🟡 In Progress | [Issue](https://github.com/DylanVann/react-native-fast-image/issues/985), [PR](https://github.com/DylanVann/react-native-fast-image/pull/951)       |
| [react-native-blur](https://github.com/Kureev/react-native-blur/)                                            | 🟢 Complete    | [v4.3.0](https://github.com/Kureev/react-native-blur/releases/tag/v4.3.0)                                                                           |

**Comment by dcalhoun** [comment]
Noting that `react-native-prompt-android` was removed in https://github.com/WordPress/gutenberg/pull/54096.

**Comment by fluiddot** [comment]
Regarding the dependencies not supporting yet the new architecture, starting in RN 0.72 we could use the feature _[New Renderer Interop Layer](https://github.com/reactwg/react-native-new-architecture/discussions/135)_ as a workaround.

> TL;DR: With React Native 0.72, we are releasing an interop layer for both Android and iOS to use non-migrated libraries on New Architecture apps.

**Comment by JerakRus** [comment]
react-native-safe-area I would recommend replacing it with
 https://github.com/th3rdwave/react-native-safe-area-context

**Comment by WoLewicki** [comment]
Hey! We at [Software Mansion](https://swmansion.com/) have been working on [improving support](https://blog.swmansion.com/sunrising-new-architecture-in-the-new-expensify-app-729d237a02f5) for the New Architecture for quite a while now. With React Native 0.76 having the New Architecture enabled by default just around the corner, it seems like a high time to migrate! Would you be interested in collaboration on that topic?

**Comment by dcalhoun** [comment]
Hey @WoLewicki! 👋🏻 Thanks for reaching out.

The mobile variant of the Gutenberg block editor is open source and under a GPL license. We welcome contributions to the code base, as well as contributions to other parts of the [WordPress project/community](https://make.wordpress.org). 

In regards to collaboration with current contributors, we agree that migrating to the New Architecture is important. However, we are currently focused on some R&D regarding the mobile editing experience at this time. We do not plan to prioritize addressing this specific issue at this time.

Please feel free to continue sharing thoughts on this issue or join the [#mobile](https://wordpress.slack.com/archives/C02RQC4LY) WordPress Slack channel. Community members are happy to discuss ideas, plans, and contributions. Thanks!

**Comment by youknowriad** [comment]
I think the mobile app development is largely stale at the moment and web view should be considered. I'm thinking of closing this issue at the moment until change of plans.
